### PR TITLE
Fix a crash when looking up an __init__ method

### LINKED ISSUE #7742: Crash: `ClassDef` has no attribute `is_abstract`
### Bug description

Code:
```python
class Foo:
    __init__ = 42


class Bar(Foo):
    def __init__(self):
        pass
```

### Command used

```shell
pylint asd.py
```

### Pylint output

```shell
************* Module asd
asd.py:8:0: C0305: Trailing newlines (trailing-newlines)
asd.py:1:0: C0114: Missing module docstring (missing-module-docstring)
asd.py:1:0: C0115: Missing class docstring (missing-class-docstring)
asd.py:1:0: R0903: Too few public methods (0/2) (too-few-public-methods)
asd.py:5:0: C0115: Missing class docstring (missing-class-docstring)
Exception on node <FunctionDef.__init__ l.6 at 0x102b8b130> in file '/Users/tusharsadhwani/code/marvin-python/asd.py'
Traceback (most recent call last):
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 90, in walk
    callback(astroid)
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/pylint/checkers/classes/class_checker.py", line 1169, in visit_functiondef
    self._check_init(node, klass)
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/pylint/checkers/classes/class_checker.py", line 2037, in _check_init
    to_call = _ancestors_to_call(klass_node)
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/pylint/checkers/classes/class_checker.py", line 2236, in _ancestors_to_call
    if init_node.is_abstract():
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 2009, in __getattr__
    return super().__getattr__(name)
  File "/Users/tusharsadhwani/code/marvin-python/venv3/lib/python3.10/site-packages/astroid/bases.py", line 134, in __getattr__
    return getattr(self._proxied, name)
AttributeError: 'ClassDef' object has no attribute 'is_abstract'
asd.py:1:0: F0002: asd.py: Fatal error while checking 'asd.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/tusharsadhwani/Library/Caches/pylint/pylint-crash-2022-11-10-16-02-44.txt'. (astroid-error)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

No crash

### Pylint version

```shell
pylint 2.15.5
astroid 2.12.12
Python 3.10.6 (main, Aug 30 2022, 04:58:14) [Clang 13.1.6 (clang-1316.0.21.2.5)]
```

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description
Fix a crash when a child class with an ``__init__`` method inherits from a parent class with an ``__init__`` class attribute.

Closes #7742

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
Merging upstream to fix primer following suggestion in https://github.com/PyCQA/pylint/issues/7738#issuecomment-1310691130

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -2240,15 +2240,17 @@ def _is_mandatory_method_param(self, node: nodes.NodeNG) -> bool:
...
     """Return a dictionary where keys are the list of base classes providing
     the queried method, and so that should/may be called from the method node.
     """
     to_call: dict[nodes.ClassDef, bases.UnboundMethod] = {}
     for base_node in klass_node.ancestors(recurs=False):
         try:
-            init_node: bases.UnboundMethod = next(base_node.igetattr(method))
+            init_node: bases.UnboundMethod = next(base_node.igetattr(method_name))
+            if not isinstance(init_node, astroid.UnboundMethod):
+                continue
```
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:2253]:
```suggestion
            init_node = next(base_node.igetattr(method_name))
            if not isinstance(init_node, astroid.UnboundMethod):
                continue
```

:smile: Apparently this typing is not correct right?

Comment (mbyrnepr2) [reply pylint/checkers/classes/class_checker.py:2253]: Was thinking the same thing. Thanks for the suggestion.

**Review by DanielNoord — DISMISSED** [review]

**Comment by jacobtylerwalls** [comment]
@mbyrnepr2 The current job on `main` needs to complete before we can start updating the PRs and rerunning jobs, FYI. :-)

**Comment by jacobtylerwalls** [comment]
I wonder if you also need to merge main.

File: doc/whatsnew/fragments/7742.other [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fix a crash when a child class with an ``__init__`` method inherits from a parent class with an ``__init__`` class attribute.
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/7742.other:1]: Should be named ``7742.bugfix`` instead imo.

File: pylint/checkers/classes/class_checker.py [hunk 3]
```diff
@@ -2240,15 +2240,17 @@ def _is_mandatory_method_param(self, node: nodes.NodeNG) -> bool:
 
 
 def _ancestors_to_call(
-    klass_node: nodes.ClassDef, method: str = "__init__"
+    klass_node: nodes.ClassDef, method_name: str = "__init__"
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:2243]: 👍 nice clarity refactor

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

**Pull request merged**